server/data/data-library/src/test/java/org/apache/james/user/lib/UsersRepositoryContract.java

```diff
@@ -514,29 +505,107 @@ default void removeUserShouldThrowWhenUserNotInRepository(TestSystem testSystem)
                 .isInstanceOf(UsersRepositoryException.class);
         }
 
-        @Test
-        default void isAdministratorShouldReturnFalseWhenNotConfigured(TestSystem testSystem) throws Exception {
-            testee().setAdministratorId(Optional.empty());
+    }
+
+    interface UsersRepositoryImplContract {
+        interface ReadOnlyContract {
+            UsersRepositoryImpl<?> testee();
 
-            assertThat(testee().isAdministrator(testSystem.admin)).isFalse();
+            @Test
+            default void isAdministratorShouldBeCaseInsensitive(TestSystem testSystem) throws Exception {
+                testee().setAdministratorId(Optional.of(testSystem.admin));
+                assertThat(testee().isAdministrator(testSystem.adminCaseVariation))
+                    .isTrue();
+            }
         }
 
-        @Test
-        default void isAdministratorShouldReturnTrueWhenConfiguredAndUserIsAdmin(TestSystem testSystem) throws Exception {
-            testee().setAdministratorId(Optional.of(testSystem.admin));
+        interface ReadWriteContract {
+            UsersRepositoryImpl<?> testee();
+
+            @Test
+            default void isAdministratorShouldReturnFalseWhenNotConfigured(TestSystem testSystem) throws Exception {
+                testee().setAdministratorId(Optional.empty());
+
+                assertThat(testee().isAdministrator(testSystem.admin)).isFalse();
+            }
+
+            @Test
+            default void isAdministratorShouldReturnTrueWhenConfiguredAndUserIsAdmin(TestSystem testSystem) throws Exception {
+                testee().setAdministratorId(Optional.of(testSystem.admin));
+
+                assertThat(testee().isAdministrator(testSystem.admin)).isTrue();
+            }
 
-            assertThat(testee().isAdministrator(testSystem.admin)).isTrue();
+            @Test
+            default void isAdministratorShouldReturnFalseWhenConfiguredAndUserIsNotAdmin(TestSystem testSystem) throws Exception {
+                testee().setAdministratorId(Optional.of(testSystem.admin));
```

Comment:
Let me suggest have two methods

```
UsersRepository testee();

UsersRepository testee(String administratorId);
```

As part of the contract definition so that you never need contracts specific to `UsersRepositoryImpl`.

Reply:
`setAdministratorId` is a private method, I can't pass it in the not-same package.

Reply:
We can pass the administrator Id via the configuration when we create the testee IMO